File: iosApp/iosApp/ViewModel/PixViewModel.swift
import Foundation
import shared

// MARK: - UI State structs

struct PixUiState {
    var step: PixFlowStep = .enterKey
    var keyInput: String = ""
    var detectedKeyType: String? = nil       // "CPF","CNPJ","Email","Phone","RandomKey","QRCode"
    var keyValidationState: KeyValidationState = .idle
    var recipient: PixRecipientDisplay? = nil
    var amountCents: Int64 = 0
    var amountFormatted: String = "R$ 0,00"
    var description: String = ""
    var isLoading: Bool = false
    var receipt: PixReceiptDisplay? = nil
    var errorMessage: String? = nil
}

enum PixFlowStep {
    case enterKey
    case confirmRecipient
    case enterAmount
    case review
    case processing
    case receipt
}

enum KeyValidationState {
    case idle
    case valid
    case invalid(String)
}

struct PixRecipientDisplay {
    let name: String
    let institution: String
    let maskedKey: String
}

struct PixReceiptDisplay {
    let e2eId: String
    let recipientName: String
    let amountFormatted: String
    let dateTimeFormatted: String
    let authenticationCode: String
}

// MARK: - ViewModel

@MainActor
final class PixViewModel: ObservableObject {
    @Published private(set) var uiState = PixUiState()

    // MARK: Tela 1: Inserir chave

    func onKeyChanged(_ input: String) {
        uiState.keyInput = input
        let detected = IOSDependencyProvider.shared.detectPixKeyType(input: input)
        uiState.detectedKeyType = detected

        guard let type = detected, !input.isEmpty else {
            uiState.keyValidationState = .idle
            return
        }
        // Enquanto digita, mostra "valid" se já passou na validação — sem mostrar erro prematuro
        let errorReason = IOSDependencyProvider.shared.validatePixKey(key: input, typeString: type)
        uiState.keyValidationState = errorReason == nil ? .valid : .idle
    }

    func onContinueFromKey() {
        guard let type = uiState.detectedKeyType else { return }
        let errorReason = IOSDependencyProvider.shared.validatePixKey(key: uiState.keyInput, typeString: type)
        if let reason = errorReason {
            uiState.keyValidationState = .invalid(reason)
            return
        }
        lookupRecipient(pixKey: uiState.keyInput)
    }

    func onScanQRCode() {
        Task {
            guard let payload = try? await IOSDependencyProvider.shared.scanPixQRCode() else { return }
            let outcome = IOSDependencyProvider.shared.parsePixQRCode(payload: payload)
            if outcome.isSuccess, let key = outcome.pixKey, let type = outcome.keyType {
                uiState.keyInput = key
                uiState.detectedKeyType = type
                uiState.keyValidationState = .valid
                if outcome.amountCents > 0 {
                    uiState.amountCents = outcome.amountCents
                    uiState.amountFormatted = formatCents(outcome.amountCents)
                }
                lookupRecipient(pixKey: key)
            } else {
                uiState.errorMessage = outcome.errorMessage ?? "QR Code inválido"
            }
        }
    }

    // MARK: Tela 2: Confirmar destinatário

    func onConfirmRecipient() {
        uiState.step = .enterAmount
    }

    func onRejectRecipient() {
        uiState.recipient = nil
        uiState.keyValidationState = .idle
        uiState.step = .enterKey
    }

    // MARK: Tela 3: Inserir valor

    func onAmountChanged(cents: Int64) {
        uiState.amountCents = cents
        uiState.amountFormatted = formatCents(cents)
    }

    func onDescriptionChanged(_ description: String) {
        uiState.description = description
    }

    func onContinueFromAmount() {
        guard uiState.amountCents > 0 else { return }
        uiState.step = .review
    }

    // MARK: Tela 4: Revisão e confirmação biométrica

    func onConfirmTransfer() {
        guard let recipient = uiState.recipient,
              let keyType = uiState.detectedKeyType else { return }

        Task {
            uiState.step = .processing
            uiState.isLoading = true

            do {
                let outcome = try await IOSDependencyProvider.shared.executePixTransfer(
                    pixKey: uiState.keyInput,
                    typeString: keyType,
                    amountCents: uiState.amountCents,
                    description: uiState.description,
                    recipientName: recipient.name,
                    recipientInstitution: recipient.institution
                )

                if outcome.isSuccess, let e2eId = outcome.e2eId {
                    let receipt = PixReceiptDisplay(
                        e2eId: e2eId,
                        recipientName: recipient.name,
                        amountFormatted: uiState.amountFormatted,
                        dateTimeFormatted: buildDateTimeFormatted(),
                        authenticationCode: String(e2eId.suffix(6)).uppercased()
                    )
                    uiState.receipt = receipt
                    uiState.step = .receipt
                } else {
                    uiState.errorMessage = outcome.errorMessage ?? "Falha na transferência"
                    uiState.step = .review
                }
            } catch {
                uiState.errorMessage = error.localizedDescription
                uiState.step = .review
            }
            uiState.isLoading = false
        }
    }

    // MARK: Utilitários

    func onDismissError() {
        uiState.errorMessage = nil
    }

    func onReset() {
        uiState = PixUiState()
    }

    // MARK: Privados

    private func lookupRecipient(pixKey: String) {
        Task {
            uiState.isLoading = true
            do {
                let outcome = try await IOSDependencyProvider.shared.lookupPixRecipient(pixKey: pixKey)
                if outcome.isSuccess,
                   let name = outcome.name,
                   let institution = outcome.institution,
                   let key = outcome.pixKey,
                   let keyType = outcome.pixKeyType {
                    uiState.recipient = PixRecipientDisplay(
                        name: name,
                        institution: institution,
                        maskedKey: maskKey(key, type: keyType)
                    )
                    uiState.step = .confirmRecipient
                } else {
                    uiState.errorMessage = outcome.errorMessage ?? "Destinatário não encontrado"
                }
            } catch {
                uiState.errorMessage = error.localizedDescription
            }
            uiState.isLoading = false
        }
    }

    private func maskKey(_ key: String, type: String) -> String {
        switch type {
        case "CPF":
            let digits = key.filter { $0.isNumber }
            guard digits.count == 11 else { return "•••.•••.•••-••" }
            let mid = String(digits.dropFirst(3).prefix(3))
            return "•••.\(mid).•••-••"
        case "CNPJ":
            return "••.•••.•••/••••-••"
        case "Email":
            let parts = key.split(separator: "@")
            if parts.count == 2 {
                let local = String(parts[0].prefix(2)) + "•••"
                return "\(local)@\(parts[1])"
            }
            return "•••@•••"
        case "Phone":
            return "+55 (••) •••••-\(key.suffix(4))"
        case "RandomKey":
            return "\(key.prefix(8))-••••-••••-••••-••••••••••••"
        default:
            return key
        }
    }

    private func formatCents(_ cents: Int64) -> String {
        let reais = cents / 100
        let centavos = cents % 100
        return String(format: "R$ %lld,%02lld", reais, centavos)
    }

    private func buildDateTimeFormatted() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd/MM/yyyy 'às' HH:mm"
        formatter.locale = Locale(identifier: "pt_BR")
        return formatter.string(from: Date())
    }
}

File: androidApp/src/main/kotlin/com/mobileone/android/viewmodel/PixViewModel.kt
package com.mobileone.android.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mobileone.shared.domain.entity.PixKeyType
import com.mobileone.shared.domain.entity.PixTransferRequest
import com.mobileone.shared.domain.error.PixError
import com.mobileone.shared.domain.usecase.DetectPixKeyTypeUseCase
import com.mobileone.shared.domain.usecase.ExecutePixTransferUseCase
import com.mobileone.shared.domain.usecase.LookupPixRecipientUseCase
import com.mobileone.shared.domain.usecase.ParsePixQRCodeUseCase
import com.mobileone.shared.domain.usecase.ValidatePixKeyUseCase
import com.mobileone.shared.feature.pix.PixKeyValidation
import com.mobileone.shared.feature.pix.PixReceipt
import com.mobileone.shared.feature.pix.PixStep
import com.mobileone.shared.feature.pix.PixTransferUiState
import com.mobileone.shared.feature.pix.RecipientDisplay
import com.mobileone.shared.security.QRCodeScanner
import com.mobileone.shared.util.CurrencyFormatter
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * ViewModel do fluxo PIX (SPEC-003). Consome os use cases do shared e expõe
 * [PixTransferUiState] via [StateFlow]. UI Compose observa e nunca contém regras de negócio.
 */
class PixViewModel(
    private val detectKeyType: DetectPixKeyTypeUseCase,
    private val validateKey: ValidatePixKeyUseCase,
    private val lookupRecipient: LookupPixRecipientUseCase,
    private val executeTransfer: ExecutePixTransferUseCase,
    private val parseQRCode: ParsePixQRCodeUseCase,
    private val qrCodeScanner: QRCodeScanner
) : ViewModel() {

    private val _uiState = MutableStateFlow(PixTransferUiState())
    val uiState: StateFlow<PixTransferUiState> = _uiState.asStateFlow()

    // ── Tela 1: Inserir chave ─────────────────────────────────────────────────

    fun onKeyChanged(input: String) {
        val detected = detectKeyType(input)
        val validation = when {
            input.isBlank() -> PixKeyValidation.Idle
            detected == null -> PixKeyValidation.Idle
            else -> {
                val result = validateKey(input, detected)
                if (result.isSuccess) PixKeyValidation.Valid
                else PixKeyValidation.Idle  // não mostrar erro enquanto ainda digita
            }
        }
        _uiState.update {
            it.copy(keyInput = input, detectedKeyType = detected, keyValidation = validation)
        }
    }

    fun onContinueFromKey() {
        val state = _uiState.value
        val keyType = state.detectedKeyType ?: return
        val validationResult = validateKey(state.keyInput, keyType)
        if (validationResult.isFailure) {
            val error = validationResult.exceptionOrNull() as? PixError.InvalidKey
            _uiState.update {
                it.copy(keyValidation = PixKeyValidation.Invalid(error?.reason ?: "Chave inválida"))
            }
            return
        }
        lookupRecipientFor(state.keyInput)
    }

    fun onScanQRCode() {
        viewModelScope.launch {
            val scanResult = qrCodeScanner.scan()
            scanResult.onSuccess { payload ->
                val parseResult = parseQRCode(payload)
                parseResult.onSuccess { qrData ->
                    _uiState.update {
                        it.copy(
                            keyInput = qrData.pixKey,
                            detectedKeyType = PixKeyType.QRCode,
                            keyValidation = PixKeyValidation.Valid,
                            amount = qrData.amountCents ?: 0L,
                            amountFormatted = CurrencyFormatter.formatBalance(qrData.amountCents ?: 0L)
                        )
                    }
                    lookupRecipientFor(qrData.pixKey)
                }.onFailure { error ->
                    _uiState.update {
                        it.copy(error = PixTransferUiState.PixError.InvalidKey(
                            (error as? PixError.QRCodeInvalid)?.reason ?: "QR Code inválido"
                        ))
                    }
                }
            }.onFailure {
                // Usuário cancelou o scan — nenhuma ação necessária
            }
        }
    }

    // ── Tela 2: Confirmar destinatário ────────────────────────────────────────

    fun onConfirmRecipient() {
        _uiState.update { it.copy(step = PixStep.EnterAmount) }
    }

    fun onRejectRecipient() {
        _uiState.update {
            it.copy(step = PixStep.EnterKey, recipient = null, keyValidation = PixKeyValidation.Idle)
        }
    }

    // ── Tela 3: Inserir valor ─────────────────────────────────────────────────

    fun onAmountChanged(cents: Long) {
        _uiState.update {
            it.copy(
                amount = cents,
                amountFormatted = CurrencyFormatter.formatBalance(cents)
            )
        }
    }

    fun onDescriptionChanged(description: String) {
        _uiState.update { it.copy(description = description) }
    }

    fun onContinueFromAmount() {
        if (_uiState.value.amount <= 0) return
        _uiState.update { it.copy(step = PixStep.Review) }
    }

    // ── Tela 4: Revisão / Confirmação ─────────────────────────────────────────

    fun onConfirmTransfer() {
        val state = _uiState.value
        val recipient = state.recipient ?: return
        val keyType = state.detectedKeyType ?: return

        viewModelScope.launch {
            _uiState.update { it.copy(step = PixStep.Processing, isLoading = true) }

            val request = PixTransferRequest(
                pixKey = state.keyInput,
                pixKeyType = keyType,
                amountCents = state.amount,
                description = state.description,
                recipientName = recipient.name,
                recipientTaxId = "",  // maskedKey no display, taxId não exposto na UI
                recipientInstitution = recipient.institution
            )

            executeTransfer(request)
                .onSuccess { e2eId ->
                    val receipt = PixReceipt(
                        transactionId = "TXN-${e2eId.takeLast(8)}",
                        e2eId = e2eId,
                        recipientName = recipient.name,
                        amountFormatted = state.amountFormatted,
                        dateTimeFormatted = buildDateTimeFormatted(),
                        authenticationCode = e2eId.takeLast(6).uppercase()
                    )
                    _uiState.update {
                        it.copy(step = PixStep.Receipt, isLoading = false, receipt = receipt)
                    }
                }
                .onFailure { error ->
                    val uiError = error.toUiError()
                    _uiState.update {
                        it.copy(
                            step = if (error is PixError.BiometricCancelled) PixStep.Review else PixStep.Review,
                            isLoading = false,
                            error = uiError
                        )
                    }
                }
        }
    }

    // ── Utilitários ───────────────────────────────────────────────────────────

    fun onDismissError() {
        _uiState.update { it.copy(error = null) }
    }

    fun onReset() {
        _uiState.value = PixTransferUiState()
    }

    private fun lookupRecipientFor(key: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            lookupRecipient(key)
                .onSuccess { recipient ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            recipient = RecipientDisplay(
                                name = recipient.name,
                                institution = recipient.institution,
                                maskedKey = maskKey(recipient.pixKey, recipient.pixKeyType)
                            ),
                            step = PixStep.ConfirmRecipient
                        )
                    }
                }
                .onFailure { error ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            error = error.toUiError()
                        )
                    }
                }
        }
    }

    private fun maskKey(key: String, type: PixKeyType): String = when (type) {
        is PixKeyType.CPF -> {
            val digits = key.filter { it.isDigit() }
            "•••.${digits.substring(3, 6)}.•••-••"
        }
        is PixKeyType.CNPJ -> {
            val digits = key.filter { it.isDigit() }
            "••.•••.${digits.substring(5, 8)}/••••-••"
        }
        is PixKeyType.Email -> {
            val parts = key.split("@")
            if (parts.size == 2) "${parts[0].take(2)}•••@${parts[1]}"
            else "•••@•••"
        }
        is PixKeyType.Phone -> "+55 (••) •••••-${key.takeLast(4)}"
        is PixKeyType.RandomKey -> "${key.take(8)}-••••-••••-••••-••••••••••••"
        is PixKeyType.QRCode -> key.take(20) + "..."
    }

    private fun buildDateTimeFormatted(): String {
        val epochSeconds = com.mobileone.shared.platform.currentEpochSeconds()
        val epochDay = (epochSeconds / 86400).toInt()
        val hour = ((epochSeconds % 86400) / 3600).toInt()
        val minute = ((epochSeconds % 3600) / 60).toInt()
        val dateStr = CurrencyFormatter.formatEpochDay(epochDay, epochDay)
        return "$dateStr às ${hour.toString().padStart(2, '0')}:${minute.toString().padStart(2, '0')}"
    }

    private fun Throwable.toUiError(): PixTransferUiState.PixError = when (this) {
        is PixError.InvalidKey -> PixTransferUiState.PixError.InvalidKey(reason)
        is PixError.RecipientNotFound -> PixTransferUiState.PixError.RecipientNotFound
        is PixError.LimitExceeded -> PixTransferUiState.PixError.LimitExceeded(
            "R$ ${limitCents / 100}"
        )
        is PixError.BiometricCancelled -> PixTransferUiState.PixError.BiometricCancelled
        is PixError.TransferFailed -> PixTransferUiState.PixError.TransferFailed(reason)
        else -> PixTransferUiState.PixError.TransferFailed(message ?: "Erro desconhecido")
    }
}
